src/multiply-matrices.js

```diff
@@ -85,3 +90,58 @@ export default function multiplyMatrices (A, B) {
 
 	return product;
 }
+
+
+// dot3 and transform functions adapted from https://github.com/texel-org/color/blob/9793c7d4d02b51f068e0f3fd37131129a4270396/src/core.js
+//
+// The MIT License (MIT)
+// Copyright (c) 2024 Matt DesLauriers
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
+// OR OTHER DEALINGS IN THE SOFTWARE.
+
+
+/**
+ * Returns the dot product of two vectors each with a length of 3.
+ *
+ * @param {Vector3} a
+ * @param {Vector3} b
+ * @returns {number}
+ */
+function dot3 (a, b) {
+	return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
+}
+
+/**
+ * Transforms a vector of length 3 by a 3x3 matrix. Specify the same input and output
+ * vector to transform in place.
+ *
+ * @param {Vector3} input
+ * @param {Matrix3x3} matrix
+ * @param {Vector3} [out]
+ * @returns {Vector3}
+*/
+export function transform (input, matrix, out = [0, 0, 0]) {
+	const x = dot3(input, matrix[0]);
+	const y = dot3(input, matrix[1]);
+	const z = dot3(input, matrix[2]);
+	out[0] = x;
+	out[1] = y;
+	out[2] = z;
```

Comment:
Presumably something like `matrix.map(m => dot3(input, m))` was slower due to the function?

Reply:
Direct access via array index is most certainly faster than a generalized iteration function. There is no doubt overhead for such convenience.

Reply:
There is some overhead to using `map` and using `map` won't allow you to transform the input in place resulting in an extra array allocation in the places where the transform is done in place (oklab color space).

I played around a bit with replacing some of the `map` calls in the RGB color spaces  with direct array access and got a modest performance improvement. Speedup for the procedural API conversion benchmark dropped to around 2.2 or 2.3 (although that might have included some optimizations to the `adapt` function, I can't remember).

Reply:
Fair enough. I suspected but figured I'd check.